### PR TITLE
Feature/esp jpeg

### PR DESCRIPTION
**Pull request by espzav**
Hi @igrr , @tore-espressif and @VojtechBartoska ,
here is first draft of JPEG component. It takes ROM code, when possible, otherwise there tjpgd inside the component.

I am thinking about this:
Add Kconfig, where can be selection if use ROM code (if possible) or not. When not used ROM code, then everything in tjpgdcnf.h should be configurable. I think, it is good for memory or speed optimalization and ROM code cannot be changed. Component is prepared for changing format there.

And please discuss the naming :-) 
Thank you very much!

### CODE REVIEW

**Comment by espzav** [comment]
One additional question for discuss: How old IDF version should be supported? I see, that for older than 4.4 should be made some changes.

**Comment by espzav** [comment]
All changes done. Reworked Kconfig for select ROM or external code, as I said in top comment.

File: components/esp_jpeg/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# JPEG Decoder: TJpgDec - Tiny JPEG Decompressor
+
+TJpgDec is a generic JPEG image decompressor module that highly optimized for small embedded systems. It works with very low memory consumption.
+
+Some microcontrollers have tjpg decoder in ROM, it is used, if there is. [^1] Using ROM code can be disabled in menuconfig.
+
+[^1]: **_NOTE:_** When there is used ROM decoder, there cannot be changed configuration.
```
Comment (tore-espressif) [inline components/esp_jpeg/README.md:7]:
```suggestion
[^1]: **_NOTE:_** When the ROM decoder is used, the configuration can't be changed. The configuration is fixed to **TODO** 
```

Please add more information about the fixed ROM configuration. + Information about how much program memory is saved by using the ROM version

File: components/esp_jpeg/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# JPEG Decoder: TJpgDec - Tiny JPEG Decompressor
+
+TJpgDec is a generic JPEG image decompressor module that highly optimized for small embedded systems. It works with very low memory consumption.
```
Comment (tore-espressif) [inline components/esp_jpeg/README.md:3]: What memory consumption can be roughly expected?
```suggestion
TJpgDec is a generic JPEG image decompressor that is highly optimized for small embedded systems. It works with very low memory consumption.
```

File: components/esp_jpeg/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+Some microcontrollers have tjpg decoder in ROM, it is used, if there is. [^1] Using ROM code can be disabled in menuconfig.
```
Comment (tore-espressif) [inline components/esp_jpeg/README.md:5]: Which MCUs exactly have the implementation in ROM?

There should be a paragraph explaining pros and cons of using the ROM implementation (flexibility, configuration, speed?, memory footprint program/data memory)

File: components/esp_jpeg/include/jpeg_decoder.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ */
+typedef struct esp_jpeg_image_output_s {
+    uint16_t width;    /*!< Width of the output image */
+    uint16_t height;   /*!< Height of the output image */
+} esp_jpeg_image_output_t;
+
+/**
+ * @brief Decode JPEG image
+ *
+ * @note This function is usually blocking.
```
Comment (tore-espressif) [inline components/esp_jpeg/include/jpeg_decoder.h:66]: Is there a case when it is not blocking?
```suggestion
 * @note This function is blocking.
```

File: components/esp_jpeg/include/jpeg_decoder.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * @param cfg: Configuration structure
+ * @param img: Output image info
+ *
+ * @return
+ *      - ESP_OK            on success
+ *      - ESP_ERR_NO_MEM    if there is no memory for allocating main structure
+ *      - ESP_FAIL          if there is an error in decoding JPEG
+ */
+esp_err_t esp_jpeg_decode(esp_jpeg_image_cfg_t *cfg, esp_jpeg_image_output_t *img);
```
Comment (tore-espressif) [inline components/esp_jpeg/include/jpeg_decoder.h:76]: could the `*cfg` be const?

File: components/esp_jpeg/jpeg_decoder.c [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+#if CONFIG_IDF_TARGET_ESP32
+#include "esp32/rom/tjpgd.h"
+#elif CONFIG_IDF_TARGET_ESP32S3
+#include "esp32s3/rom/tjpgd.h"
+#elif CONFIG_IDF_TARGET_ESP32C2
+#include "esp32c2/rom/tjpgd.h"
+#elif CONFIG_IDF_TARGET_ESP32C3
+#include "esp32c3/rom/tjpgd.h"
+#elif CONFIG_IDF_TARGET_ESP32H2
+#include "esp32h2/rom/tjpgd.h"
```
Comment (tore-espressif) [inline components/esp_jpeg/jpeg_decoder.c:30]: `#include "rom/tjpgd.h"` should work here

**Review by tore-espressif — COMMENTED** [review]
Thank you @espzav . I'll test in on HW soon

File: components/esp_jpeg/idf_component.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+version: "0.9.0"
+description: TJpgDec - Tiny JPEG Decompressor
+url: https://github.com/espressif/esp-bsp/tree/master/components/esp_jpeg/
+dependencies:
+  idf: ">=4.4"
```
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/idf_component.yml:5]: There cannot be used esp_check.h if we want to allow for IDF v4.3

File: components/esp_jpeg/Kconfig [hunk 8]
```diff
@@ -0,0 +1,66 @@
+menu "JPEG Decoder"
+
+    config JD_USE_EXT
+        bool "Use external TinyJPG Decoder instead of ROM code."
```
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/Kconfig:4]: 1. It sound good, but I don't know, how to use the `depends on JD_USE_EXT` in inverse logic... 
2. Dots will be removed.
3. I will add help.

[the remark this answers is not in the document]
Comment (tore-espressif) [reply components/esp_jpeg/Kconfig:4]: `depends on !JD_USE_EXT`

File: components/esp_jpeg/include/jpeg_decoder.h (hunk 5 above)
Comment (espzav) [reply components/esp_jpeg/include/jpeg_decoder.h:76]: It cannot be const, there is saved how many bytes was read (internally).

File: components/esp_jpeg/include/jpeg_decoder.h (hunk 4 above)
Comment (espzav) [reply components/esp_jpeg/include/jpeg_decoder.h:66]: No, there isn't.

File: components/esp_jpeg/license.txt [hunk 9]
```diff
@@ -0,0 +1,202 @@
+
+                                 Apache License
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
```
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/license.txt:4]: I don't know, which license should be there. But there is same code as in ROM, but only can be configured.

File: components/esp_jpeg/Kconfig (hunk 8 above)
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/Kconfig]: Thanks!

**Comment by espzav** [comment]
@tore-espressif Thank you for review. I added some measurement into readme and more information.

File: components/esp_jpeg/license.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/license.txt:4]: @igrr please, could you help me with license?

File: components/esp_jpeg/jpeg_decoder.c [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    /* Size of output image */
+    uint32_t outsize = (JDEC.height / scale_div) * (JDEC.width / scale_div) * out_color_bytes;
+    ESP_GOTO_ON_FALSE((outsize <= cfg->outbuf_size), ESP_ERR_NO_MEM, err, TAG, "Not enough size in output buffer!");
+
+    /* Size of output image */
+    img->height = JDEC.height / scale_div;
+    img->width = JDEC.width / scale_div;
+
+    /* Decode JPEG */
+    res = jd_decomp(&JDEC, jpeg_decode_out_cb, cfg->out_scale);
```
Comment (tore-espressif) [inline components/esp_jpeg/jpeg_decoder.c:97]: I'm getting some compilation warnings:
```
C:/github_repos/esp-bsp/components/esp_jpeg/jpeg_decoder.c:97:28: warning: passing argument 2 of 'jd_decomp' from incompatible pointer type [-Wincompatible-pointer-types]
     res = jd_decomp(&JDEC, jpeg_decode_out_cb, cfg->out_scale);
                            ^~~~~~~~~~~~~~~~~~
In file included from C:/github_repos/esp-bsp/components/esp_jpeg/jpeg_decoder.c:25:
C:/github_repos/esp-bsp/components/esp_jpeg/tjpgd/tjpgd.h:95:36: note: expected 'int (*)(JDEC *, void *, JRECT *)' {aka 'int (*)(struct JDEC *, void *, struct <anonymous> *)'} but argument is of type 'unsigned int (*)(JDEC *, void *, JRECT *)' {aka 'unsigned int (*)(struct JDEC *, void *, struct <anonymous> *)'}
 JRESULT jd_decomp (JDEC *jd, int (*outfunc)(JDEC *, void *, JRECT *), uint8_t scale);
                              ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

**Review by tore-espressif — CHANGES_REQUESTED** [review]

File: components/esp_jpeg/jpeg_decoder.c [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+#include "esp_system.h"
+#include "esp_rom_caps.h"
+#include "esp_log.h"
+#include "esp_err.h"
+#include "esp_check.h"
+#include "jpeg_decoder.h"
+
+#if CONFIG_JD_USE_ROM
+/* When supported in ROM, use ROM functions */
+#if (CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32S3 || CONFIG_IDF_TARGET_ESP32C2 || CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32H2)
```
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/jpeg_decoder.c:18]: I know about it, but it is not supported in IDF 4.4 and older.

File: components/esp_jpeg/jpeg_decoder.c (hunk 10 above)
Comment (espzav) [reply components/esp_jpeg/jpeg_decoder.c:97]: The TJPG Decoder in ROM is little different than TJPG decoder code in component. There is newer version in component. I didn't want to add more #if-def for this warning.

File: components/esp_jpeg/README.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+- ESP32-C2
+- ESP32-C3
+- ESP32-H2
+
+### Fixed compilation configuration of the ROM code
+- Stream input buffer: 512
+- Output pixel format: RGB888
+- Descaling feature for output: Enabled
+- Table for saturation: Enabled
+- Optimization level: Basic
```
[the remark this answers is not in the document]
Comment (espzav) [reply components/esp_jpeg/README.md:39]: It equals to (JD_FASTDECODE = 0)

File: components/esp_jpeg/Kconfig [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            depends on !JD_USE_ROM
+            default JD_FORMAT_RGB888
+            help
+                Output format of the TJpg decoder. The component can select the output format in runtime,
+                but it depends on this choice.
+
+        config JD_FORMAT_RGB888
+            bool "RGB888 (24-bit/pix)"
+        config JD_FORMAT_RGB565
+            bool "RGB565 (16-bit/pix)"
```
Comment (tore-espressif) [inline components/esp_jpeg/Kconfig:34]: Based on our discussion, should this be:
```suggestion
        config JD_FORMAT_RGB888
            bool "RGB565 and RGB888 (24-bit/pix)"
        config JD_FORMAT_RGB565
            bool "RGB565 (16-bit/pix)"
```
?

**Review by tore-espressif — APPROVED** [review]
One minor comment, otherwise LGTM!

File: components/esp_jpeg/Kconfig (hunk 13 above)
Comment (espzav) [reply components/esp_jpeg/Kconfig:34]: I don't know, what is the right. I thought, that it is only for settings TJPG decoder code. It set only output of the TJPG. Output of whole component is depend on runtime option.

**Comment by espzav** [comment]
@tore-espressif Thank you! But I still don't know, what the license should I use.

File: components/esp_jpeg/Kconfig (hunk 13 above)
Comment (tore-espressif) [reply components/esp_jpeg/Kconfig:34]: I see. Could the labels be "Support RGB565 output" and "Support RGB565 and RGB888 output" 

and in help we could state that "Output format is selected at runtime"

?

**Pull request merged**